pkg/agent/apiserver/handlers/bundle/handler.go

```diff
@@ -0,0 +1,336 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package bundle
+
+import (
+	"archive/tar"
+	"crypto/sha256"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"os/exec"
+	"path"
+	"path/filepath"
+	"strings"
+	"sync"
+	"time"
+
+	"k8s.io/klog"
+
+	systemv1beta1 "github.com/vmware-tanzu/antrea/pkg/apis/system/v1beta1"
+)
+
+type handler struct {
+	sync.Mutex
+	forceRefresh chan struct{}
+	cache        Response
+}
+
+func (h *handler) clean(duration time.Duration) {
+	select {
+	case <-h.forceRefresh:
+		return
+	case <-time.After(duration):
+	}
+	h.removeCache()
+}
+
+type CollectRequest struct {
+	CollectItems []string `json:"collectItems"`
+	Force        bool     `json:"force"`
+}
+
+type Response struct {
+	Status     string `json:"status"`
+	Sum        string `json:"sum,omitempty"`
+	SubmitTime int64  `json:"generateTime,omitempty"`
+	ExpireTime int64  `json:"expireTime,omitempty"`
+	FilePath   string `json:"filePath,omitempty"`
+}
+
+func dumpFlows(basedir string) error {
+	brListOutput, err := exec.Command("ovs-vsctl", "list-br").Output()
+	if err != nil {
+		return fmt.Errorf("error when collecting ovs bridge info: %w", err)
+	}
+	err = os.Mkdir(filepath.Join(basedir, "flows"), os.ModePerm)
+	if err != nil {
+		return fmt.Errorf("error when creating flows output dir: %w", err)
+	}
+	for _, brName := range strings.Split(strings.TrimSpace(string(brListOutput)), "\n") {
+		dumpFlowOutput, err := exec.Command("ovs-ofctl", "dump-flows", brName).Output()
+		if err != nil {
+			return fmt.Errorf("error when dumping flows on bridge %s: %w", brName, err)
+		}
+		err = ioutil.WriteFile(filepath.Join(basedir, "flows", brName), dumpFlowOutput, 0644)
+		if err != nil {
+			return fmt.Errorf("error when creating flows output file: %w", err)
+		}
+	}
+	return nil
+}
+
+func dumpIPTables(basedir string) error {
+	iptablesOutput, err := exec.Command("iptables-save").Output()
+	if err != nil {
+		return fmt.Errorf("error when dumping iptables data: %w", err)
+	}
+	err = ioutil.WriteFile(filepath.Join(basedir, "iptables"), iptablesOutput, 0644)
+	if err != nil {
+		return fmt.Errorf("error when writing iptables dumps: %w", err)
+	}
+	return nil
+}
+
+func dumpAgentLog(basedir string) error {
+	targetDir := filepath.Join(basedir, "logs")
+	err := os.Mkdir(targetDir, os.ModePerm)
+	if err != nil {
+		return fmt.Errorf("error when creating logs output dir: %w", err)
+	}
+	logDir := "/var/log/antrea"
+	return filepath.Walk(logDir, func(filePath string, info os.FileInfo, err error) error {
```

Comment:
Do you think we can enable some code sharing for all these dumpXXXLog() funcs in agent and controller?

Reply:
Yes, I also considered this but I'm not sure where should I put the log dump function in `pkg/controller/util/dump.go` and `pkg/agent/util/dump_linux.go`.

Reply:
How about  pkg/util/dump.go?